### ISSUE TITLE
AlignerNet instead of MAS

### ISSUE THREAD

**Issue by codeghees** [p0p4k/vits2_pytorch#81]
Is it possible to use AlignerNet (aligner.py in pflow-tts repo) instead of MAS in VITS2? 

What should be changed in the code? I am a bit confused on what the inputs should be.

**Comment by p0p4k** [comment]
Check out pflow repo for guidance.

**Comment by codeghees** [comment]
Hey, thanks for replying! 

I did take most of the code from that repo. I am trying to debug why my alignment curve looks like this:
![image](https://github.com/p0p4k/vits2_pytorch/assets/40551128/c55c156f-e367-4e77-98fc-864653c3b49c)

`align_loss` is being added to duration loss.
Inputs: 
```
 aln_hard, aln_soft, aln_log, aln_mask = self.aligner(
            m_p.transpose(1,2), x_mask, y, y_mask
            )
attn = aln_mask.transpose(1,2).unsqueeze(1)
align_loss = self.aligner_loss(aln_log, x_lengths, y_lengths)
m_p is returned by TextEncoder.
```
Appreciate any insights!

**Comment by p0p4k** [comment]
It might be that it is still learning the durations. It think MAS is good enough. What is important is the duration_predictor module.

**Comment by codeghees** [comment]
So this above graph is after training for several days and thousands of steps. It seems like some bug - maybe in shape size or sth.

The output is basically the first word is legible but the rest is basically gibberish.

**Comment by p0p4k** [comment]
I see. Maybe it needs some fixing to do, if you can start a PR we can debug this together. I am busy with other stuff.

**Comment by Tera2Space** [comment]
Probably same problem as we have in pflow)

**Comment by codeghees** [comment]
@Tera2Space What was the problem?

**Comment by Tera2Space** [comment]
> @Tera2Space What was the problem?

Code geass nice, problem was that it generate wrong aligment, i just though of reason: https://github.com/p0p4k/pflowtts_pytorch/issues/24#issuecomment-1983667975

in your model, what shape is input to alignernet?

**Comment by codeghees** [comment]
@Tera2Space @p0p4k I added a basic PR of the changes I have so far: https://github.com/p0p4k/vits2_pytorch/pull/82

**Comment by codeghees** [comment]
For a single batch - shapes look something like this:
```
m_p torch.Size([32, 192, 74])
x_mask torch.Size([32, 1, 74])
y_mask torch.Size([32, 1, 293])
m_p torch.Size([32, 192, 74])
x torch.Size([32, 192, 74])
y torch.Size([32, 80, 293])
m_p.transpose(1,2) torch.Size([32, 74, 192])
```

**Comment by Tera2Space** [comment]
> For a single batch - shapes look something like this:
> 
> ```
> m_p torch.Size([32, 192, 74])
> x_mask torch.Size([32, 1, 74])
> y_mask torch.Size([32, 1, 293])
> m_p torch.Size([32, 192, 74])
> x torch.Size([32, 192, 74])
> y torch.Size([32, 80, 293])
> m_p.transpose(1,2) torch.Size([32, 74, 192])
> ```

hm than my guess was wrong/there are more problems. 
I tested my idea on pflow and it didn't work too, so probably we need to check for another problems, I think we should use something like https://github.com/lucidrains/naturalspeech2-pytorch as reference.

**Comment by p0p4k** [comment]
Iirc, I had yanked the alignernet from there.

**Comment by codeghees** [comment]
Yeah currently trying out various inputs to Aligner; possibly an input issue. It might be an issue with the masks.

**Comment by p0p4k** [comment]
still not convinced about putting efforts in aligner net, we must focus on a better `TextEncoder` instead.

**Comment by codeghees** [comment]
do you think that's a bottleneck right now?

**Comment by p0p4k** [comment]
For vits2 it should be the duration predictor, and for pflow it should be both textencoder and duration predictor. MAS gives  good alignments during training, it is during inference that these models perform worse.

**Comment by Tera2Space** [comment]
> pflow it should be both textencoder

how do you think we can improve pflow's encoder?

**Comment by nicemanis** [comment]
> Hey, thanks for replying!
> 
> I did take most of the code from that repo. I am trying to debug why my alignment curve looks like this: ![image](https://private-user-images.githubusercontent.com/40551128/310711805-c55c156f-e367-4e77-98fc-864653c3b49c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk4MTUyMTksIm5iZiI6MTcxOTgxNDkxOSwicGF0aCI6Ii80MDU1MTEyOC8zMTA3MTE4MDUtYzU1YzE1NmYtZTM2Ny00ZTc3LTk4ZmMtODY0NjUzYzNiNDljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAxVDA2MjE1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0OWE2MTNmNWNmYWQ4YWNmZjZiOGYxOTY2NWIyYWRjMGNkNGU0YTcyZDExZDViZWU2NjQxYjE5ZmI5MDk3Y2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9wFLnLiaBRUqZJDv6vj_-QFyqqaH3EVRS8YsDHzpQbw)
> 
> `align_loss` is being added to duration loss. Inputs:
> 
> ```
>  aln_hard, aln_soft, aln_log, aln_mask = self.aligner(
>             m_p.transpose(1,2), x_mask, y, y_mask
>             )
> attn = aln_mask.transpose(1,2).unsqueeze(1)
> align_loss = self.aligner_loss(aln_log, x_lengths, y_lengths)
> m_p is returned by TextEncoder.
> ```
> 
> Appreciate any insights!

Is padding included in the encoder timesteps (it seems to me that it is)? You can remove the padded part from the plot.

**Comment by brunopiovan** [comment]
I know this is a bit old but I decided to play with this a bit and I think I managed to get something to work...

<img width="1836" height="906" alt="Image" src="https://github.com/user-attachments/assets/eaf532f4-176e-48c5-96fe-564650846713" />

**Comment by JohnHerry** [comment]
Is AlignerNet get better duration distribution then?
